src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -1394,6 +1438,72 @@ private void WritePackageFolders()
+                foreach (var package in _lockFile.Libraries)
+                {
+                    var packageName = package.Name;
+                    var packageVersion = package.Version.ToNormalizedString();
+                    string packageId = $"{packageName}/{packageVersion}";
+
+                    WriteItem(packageId);
+                    WriteMetadata(MetadataKeys.Name, packageName);
+                    WriteMetadata(MetadataKeys.Type, package.Type);
+                    WriteMetadata(MetadataKeys.Version, packageVersion);
+                    WriteMetadata(MetadataKeys.Path, package.Path ?? string.Empty);
+
+                    string resolvedPackagePath = ResolvePackagePath(package);
+                    WriteMetadata(MetadataKeys.ResolvedPath, resolvedPackagePath ?? string.Empty);
+
+                    WriteMetadata(MetadataKeys.DiagnosticLevel, GetPackageDiagnosticLevel(package));
+                }
+
+                string ResolvePackagePath(LockFileLibrary package)
+                {
+                    if (package.IsProject())
+                    {
+                        var relativeMSBuildProjectPath = package.MSBuildProject;
+
+                        if (string.IsNullOrEmpty(relativeMSBuildProjectPath))
+                        {
+                            throw new BuildErrorException(Strings.ProjectAssetsConsumedWithoutMSBuildProjectPath, package.Name, _task.ProjectAssetsFile);
+                        }
+
+                        return GetAbsolutePathFromProjectRelativePath(relativeMSBuildProjectPath);
+                    }
+                    else
+                    {
+                        return _packageResolver.GetPackageDirectory(package.Name, package.Version);
+                    }
+                }
+
+                string GetAbsolutePathFromProjectRelativePath(string path)
+                {
+                    return Path.GetFullPath(Path.Combine(Path.GetDirectoryName(_task.ProjectPath), path));
+                }
+
+                string GetPackageDiagnosticLevel(LockFileLibrary package)
+                {
+                    string target = _task.TargetFramework ?? "";
+
+                    var messages = _lockFile.LogMessages.Where(log => log.LibraryId == package.Name && log.TargetGraphs
+                                    .Select(tg =>
+                                    {
+                                        var parsedTargetGraph = NuGetFramework.Parse(tg);
+                                        var alias = _lockFile.PackageSpec.TargetFrameworks.FirstOrDefault(tf => tf.FrameworkName == parsedTargetGraph)?.TargetAlias;
+                                        return alias ?? tg;
+                                    }).Contains(target));
+
+                    if (messages.Count() == 0)
+                    {
+                        return string.Empty;
+                    }
+
+                    return messages.Max(log => log.Level).ToString();
```

Comment:
This is double-enumerating `messages`. Use a `foreach` here to ensure items are only enumerated once.